django/webapp/views.py

```diff
@@ -1,15 +1,128 @@
-from django.views.generic import DetailView, ListView
+from django.views.generic import DetailView, ListView, View
+from django.http import HttpResponse
+from django.contrib.auth import logout
+from django.contrib.auth.mixins import UserPassesTestMixin
+from django.core.exceptions import SuspiciousOperation
+from django.contrib.auth.forms import PasswordResetForm
 
-from webapp.models import Synagogue, Member
+from webapp.models import Synagogue, Member, get_from_model
+from webapp.forms import AddSynagogueForm, LoginForm, AddUserToSynagogueForm
+
+import abc
+
+
+class SynagoguePermChecker(UserPassesTestMixin):
+    """
+    an abstract class to check if the user has admin permissions for the requested synagogue.
+    every child class need to implement get_synagogue which get the synagogue we are trying to query about
+    from the request. It will return 403 if he is unauthorized.
+    This class should be the first to inherit from.
+    """
+    raise_exception = True
+
+    def test_func(self):
+        synagogue = self.get_synagogue()
+        return synagogue.admins in self.request.user.groups.all()
+
+    @abc.abstractmethod
+    def get_synagogue(self):
+        pass
+
+
+class MyFormView(View):
+    """
+    django's FormView is not what we need since it does a lot that we don't need (e.g implement a GET that render
+    template, but we only want a simple post). So this simple FormView is exactly what we need. it is separated to
+    a lot of functions to let the children hook it. The usual child should just add
+    form = <MyForm>
+    and let the form do all the job.
+    """
+    form = NotImplementedError
+
+    def post(self, request):
+        form = self.get_form()
+        self.save(form)
+        return HttpResponse()
+
+    def get_form(self):
+        # we validate the form in get_form since the form is useless before it is validated
+        form = self.form(self.request.POST)
+        if not self.validate(form):
+            raise SuspiciousOperation()
+        return form
+
+    def validate(self, form):
+        return form.is_valid()
+
+    def save(self, form, *args, **kwargs):
+        form.save(*args, **kwargs)
+
+
+class SafeGetMixin:
```

Comment:
this seems a bit over-engineered. is it really worth the mixin rather than just having a function which takes the kwargs dict as a parameter?

similarly to get_from_model, I'm not sure it's generic enough to be useful in many different situations.